### ISSUE TITLE
std::vector<Vector3f> to NumPy?

### ISSUE THREAD

**Issue by vadimcn** [wlav/cppyy#183]
What is the most straightforward way of converting `std::vector<Vector3f>` to a NumPy array?  (where Vector3f is effectively `struct { float x, y, z; }`)

Also, please enable "Discussions" for this repo :-)

**Comment by wlav** [comment]
No, as there's no connection between the two structures. However, I don't think that struct will see any padding (you can instruct the compiler to pack it to be sure), so you could write a C++ helper, relying on the fact that at that point, the memory layout will be contiguous, so a `memcpy` can be used.

I though "Discussions" was active. I just clicked on a first post, as maybe that was needed on top of the check mark.

**Comment by vadimcn** [comment]
Floats are 4-byte aligned, so there shouldn't be any padding for Vector3f...
But anyhow, assuming I can figure out layout and alignment, how do I get a Python buffer object for contents of my std::vector?   I was going try [this](https://cppyy.readthedocs.io/en/latest/lowlevel.html#numpy-casts), however `my_vector.data()` seems to return the first array element, not a `LowLevelView` object.

**Comment by wlav** [comment]
Yes, in C++ it's not possible to distinguish between a pointer to an object and a pointer to an array of objects. Since the former is vastly more common, it's the default. I should probably change the pythonization of `std::vector<>::data` to tell it that an array is returned.

Anyway, you can still force it to be an array by reshaping the pointer.

I was thinking of using a C++ helper to be able to do the `memcpy`. Can stay in Python, I guess, by using `cppyy.ll.cast`, although to be sure `ll.cast` is a C++ helper. :)

Here's some code that may be useful (the last steps are just there to prove that no copying is happening):

```
import cppyy
import cppyy.ll
import numpy as np

cppyy.cppdef("""\
struct __attribute__((__packed__)) Vector3f
{
   float x, y, z;
};""")

N = 5

v = cppyy.gbl.std.vector['Vector3f'](N)

for i in range(N):
    d = v[i]
    d.x, d.y, d.z = i, i*N, i*N**2

data = v.data()
data.__reshape__((len(v),))

for i in range(N):
    d = data[i]
    print(d.x, d.y, d.z)

buf = cppyy.ll.cast['void*'](cppyy.addressof(v.data()))
nv = np.frombuffer(buf, dtype=np.float32, count=3*N)
nv = nv.reshape((N, 3))

for i in range(N):
    print(nv[i, 0], nv[i, 1], nv[i, 2])

for i in range(N):
    d = v[i]
    d.x, d.y, d.z = d.z, d.x, d.y

for i in range(N):
    print(nv[i, 0], nv[i, 1], nv[i, 2])
```

Not particularly clean, but that can be hidden in a pythonization by adding the cast and `frombuffer` code in a method `to_numpy` that is added to the `Vector3f` class. Probably not too efficient either, but likely still faster than copying large arrays.

**Comment by vadimcn** [comment]
That works, thanks!

Just curious, was there no way to directly get a buffer object from the `data` array after this:
```python
data = v.data()
data.__reshape__((len(v),))
```

**Comment by wlav** [comment]
The basic issue is that `data` is still a `CPPInstance`. The reason is that both a pointer to an object and a pointer to an array are the same type in C++ (`Vector3f*` in both cases here) and therefore represented the same way in the bindings.

For the fun of it, I implemented the pythonization that makes `vector<Vector3f>::data()` return an array (incl. the proper size). That works fine (and is in repo), but then to be able to pass it directly to `np.frombuffer`, it needs to implement something that numpy understands, e.g. the buffer protocol. Fine for the example here, but it breaks tons of tests. A recurring reason seems to be that Python prefers the buffer protocol over the iterator protocol and so loops over containers become loops over their byte representation rather than their payload.

So in short, yes, it can be done, but no, an implicit cast to `void*` appears to be massive trouble. (Not at all surprising, of course.)

An alternative approach would be to implement a `cppyy.ll.as_memoryview()` (alongside several other existing methods such as `as_cobject()`, `as_capsule()`, and `as_ctypes()` for rather similar purposes, but none of which numpy accepts). It would return a builtin `memoryview` object, which supports a byte-wise buffer protocol. That would already be cleaner (and more efficient) than the code above.

**Comment by wlav** [comment]
I added `cppyy.ll.as_memoryview()` as it's simple enough. For now, it rejects all but arrays of aggregates of known size, which covers the use case here. So now this works:
```
nv = np.frombuffer(cppyy.ll.as_memoryview(v.data()), dtype=np.float32)
```
Notice how there are no sizes specified anymore (the resulting numpy view will still be 1-D, so reshaping it like in the example is still needed: there's no knowledge of `Vector3f` being 3 floats being passed around).
